Replace CXF with Jersey2 [ZEPPELIN-903]

echarles opened the pull request:
### What is this PR for?

Use jersey2 instead of CXF to manage JSON REST requests and responses.
Jersey2 can better manage the JSON bodies and mappings, mapping the json attributes to a POJO (getter/setter) being given as parameter of the method.
### What type of PR is it?

[Improvement]
### Todos
### What is the Jira issue?
- https://issues.apache.org/jira/browse/ZEPPELIN-903
### How should this be tested?

Build, run and check the 
### Screenshots (if appropriate)

N/A
### Questions:
- Does the licenses files need update? Y
- Is there breaking changes for older versions? N
- Does this needs documentation? N

echarles commented:
Any feedback on moving to jersey 2? (btw CI is green)

Leemoonsoo commented:
Thanks for the contribution.

Could you take care of zeppelin-distribution/src/bin_license/LICENSE for dependency change?

I have summarized cxf and jersey related dependencies (transitive dependencies).
cxf related entry need to be removed from `bin_license/LICENSE` and jersey related entry need to be updated/added to `bin_license/LICENSE`.

Before
```
[INFO] +- org.apache.cxf:cxf-rt-frontend-jaxrs:jar:2.7.7:compile
[INFO] |  +- org.apache.cxf:cxf-rt-core:jar:2.7.7:compile
[INFO] |  |  \- com.sun.xml.bind:jaxb-impl:jar:2.2.6:compile
[INFO] |  \- org.apache.cxf:cxf-rt-bindings-xml:jar:2.7.7:compile

[INFO] +- org.apache.cxf:cxf-rt-transports-http:jar:2.7.7:compile

[INFO] +- org.apache.cxf:cxf-rt-transports-http-jetty:jar:2.7.7:compile
[INFO] |  \- org.apache.geronimo.specs:geronimo-servlet_3.0_spec:jar:1.0:compile

[INFO] +- org.apache.cxf:cxf-api:jar:2.7.7:compile
[INFO] |  +- org.codehaus.woodstox:woodstox-core-asl:jar:4.2.0:compile
[INFO] |  |  \- org.codehaus.woodstox:stax2-api:jar:3.1.1:compile
[INFO] |  +- org.apache.ws.xmlschema:xmlschema-core:jar:2.0.3:compile
[INFO] |  +- org.apache.geronimo.specs:geronimo-javamail_1.4_spec:jar:1.7.1:compile
[INFO] |  \- wsdl4j:wsdl4j:jar:1.6.3:compile

[INFO] +- com.sun.jersey:jersey-servlet:jar:1.13:compile
[INFO] |  \- com.sun.jersey:jersey-server:jar:1.13:compile
[INFO] |     +- asm:asm:jar:3.1:compile
[INFO] |     \- com.sun.jersey:jersey-core:jar:1.13:compile
```

After
```
[INFO] +- org.glassfish.jersey.core:jersey-client:jar:2.22.2:compile
[INFO] |  +- javax.ws.rs:javax.ws.rs-api:jar:2.0.1:compile
[INFO] |  +- org.glassfish.jersey.core:jersey-common:jar:2.22.2:compile
[INFO] |  |  +- org.glassfish.jersey.bundles.repackaged:jersey-guava:jar:2.22.2:compile
[INFO] |  |  \- org.glassfish.hk2:osgi-resource-locator:jar:1.0.1:compile
[INFO] |  +- org.glassfish.hk2:hk2-api:jar:2.4.0-b34:compile
[INFO] |  |  +- org.glassfish.hk2:hk2-utils:jar:2.4.0-b34:compile
[INFO] |  |  \- org.glassfish.hk2.external:aopalliance-repackaged:jar:2.4.0-b34:compile
[INFO] |  +- org.glassfish.hk2.external:javax.inject:jar:2.4.0-b34:compile
[INFO] |  \- org.glassfish.hk2:hk2-locator:jar:2.4.0-b34:compile
[INFO] |     \- org.javassist:javassist:jar:3.18.1-GA:compile

[INFO] +- org.glassfish.jersey.core:jersey-server:jar:2.22.2:compile
[INFO] |  +- org.glassfish.jersey.media:jersey-media-jaxb:jar:2.22.2:compile
[INFO] |  +- javax.annotation:javax.annotation-api:jar:1.2:compile
[INFO] |  \- javax.validation:validation-api:jar:1.1.0.Final:compile

[INFO] +- org.glassfish.jersey.containers:jersey-container-servlet-core:jar:2.22.2:compile

[INFO] +- org.glassfish.jersey.media:jersey-media-json-jackson:jar:2.22.2:compile
[INFO] |  +- org.glassfish.jersey.ext:jersey-entity-filtering:jar:2.22.2:compile
[INFO] |  +- com.fasterxml.jackson.jaxrs:jackson-jaxrs-base:jar:2.5.4:compile
[INFO] |  |  +- com.fasterxml.jackson.core:jackson-core:jar:2.5.4:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-databind:jar:2.5.4:compile
[INFO] |  \- com.fasterxml.jackson.jaxrs:jackson-jaxrs-json-provider:jar:2.5.4:compile
[INFO] |     \- com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.5.4:compile
```

echarles commented:
@Leemoonsoo  Just pushed the asked changes regarding licenses.

echarles commented:
Failing with 

org.apache.zeppelin.interpreter.InterpreterException: org.apache.thrift.TApplicationException: Internal error processing createInterpreter
	at org.apache.zeppelin.interpreter.remote.RemoteInterpreter.init(RemoteInterpreter.java:238)

Is this a WKZCIF (Well Know Zeppelin CI Failure) ?

bzz commented:
A quick search over JIRA suggests that the failure you reffer was taked care of in [ZEPPELIN-1264](https://issues.apache.org/jira/browse/ZEPPELIN-1264) and is resolved in latest master, please rebase.


In PR body 

> Does the licenses files need update? N

I think it does

@echarles all transitive dependencies lie `woodstox` and `org.glassfish.jersey.bundles` and `javax.validation`, etc  will be included in convenience binary for release so each of them must be as well included in `zeppelin-distribution/src/bin-license/LICENSE` according to their license. Could you please double check that?

----

>Jersey2 can better manage the JSON bodies and mappings.

Could you please elaborate on that part? People might not be very familiar with it very well, so if you could put the benefits of such change in PR description - that would help a lot.

Thank you!

echarles commented:
@bzz I have updated the PR descrption (first comment) with the need to update license and adding a note on how easy it is with jersey to map a full json to a pojo being passed as method parameter.

I have also checked the transitive deps [1] and updated `zeppelin-distribution/src/bin-license/LICENSE` accordinly.

Lets's if build is green.

+ com.fasterxml.jackson.core:jackson-core:jar:2.5.4:compile
+ com.fasterxml.jackson.jaxrs:jackson-jaxrs-base:jar:2.5.4:compile
+ com.fasterxml.jackson.jaxrs:jackson-jaxrs-json-provider:jar:2.5.4:compile
+ com.fasterxml.jackson.core:jackson-databind:jar:2.5.4:compile
+ com.fasterxml.jackson.module:jackson-module-jaxb-annotations:jar:2.5.4:co

→ Apache 2

+ javax.annotation:javax.annotation-api:jar:1.2:compile

→ CDDL2

+ javax.ws.rs:javax.ws.rs-api:jar:2.0.1:compile

→ CDDL2

+ javax.validation:validation-api:jar:1.1.0.Final:compile

→ Apache2

+ org.glassfish.hk2:hk2-api:jar:2.4.0-b34:compile
+ org.glassfish.hk2:hk2-locator:jar:2.4.0-b34:compile
+ org.glassfish.hk2:hk2-utils:jar:2.4.0-b34:compile
+ org.glassfish.hk2:osgi-resource-locator:jar:1.0.1:compile
+ org.glassfish.hk2.external:aopalliance-repackaged:jar:2.4.0-b34:compile
+ org.glassfish.hk2.external:javax.inject:jar:2.4.0-b34:compile

→ CDDL2

+ org.glassfish.jersey.bundles.repackaged:jersey-guava:jar:2.22.2:compile
+ org.glassfish.jersey.containers:jersey-container-servlet-core:jar:2.22.2:compile
+ org.glassfish.jersey.core:jersey-client:jar:2.22.2:compile
+ org.glassfish.jersey.core:jersey-common:jar:2.22.2:compile
+ org.glassfish.jersey.core:jersey-server:jar:2.22.2:compile
+ org.glassfish.jersey.ext:jersey-entity-filtering:jar:2.22.2:compile
+ org.glassfish.jersey.media:jersey-media-jaxb:jar:2.22.2:compile
+ org.glassfish.jersey.media:jersey-media-json-jackson:jar:2.22.2:compile

→ CDDL

+ org.javassist:javassist:jar:3.18.1-GA:compile

→ Apache2

bzz commented:
@echarles thank you for taking care! 
Please give me some time to try it out and make another pass on it and I'll get back to you here.

bzz commented:
Thank you for your patience, @echarles 

AFAIK CDDL 1.1 that jersey2 is distributed under belongs to ASF [Category B license list](http://www.apache.org/legal/resolved.html#category-b) so it can only be included in ASF projects in binary form (that's what this PR is doing), so this makes it not required to include license text to `./license/*` (reserved for source code), but only to `zeppelin-distribution/src/bin_license/`.

Could you please remove full text of the licence from `./license/*` ?

Another thing is, as one can see, this list of licenses only includes
> Common Development and Distribution Licenses: CDDL, Versions 1.0 and 1.1 

but there are components under CDDL 2.0:

```
(CDDL 2.0) javax.ws.rs-api (javax.ws.rs - https://jax-rs-spec.java.net/)
(CDDL 2.0) validation-api (javax.validation - http://beanvalidation.org/)
(CDDL 2.0) javax.annotation-api (javax.annotation - https://www.jcp.org/en/jsr/detail?id=250)
```

I will look more into that, meanwhile could you please include full text of this license to `zeppelin-distribution/src/bin_license/` as well?

echarles commented:
@bzz I have moved  `LICENSE-jersery-2` to `zeppelin-distribution/src/bin_license`.

echarles commented:
@bzz  are we ok now with the license information?

Travis is red for 3 on the 9 profiles, but this comes from a timeout trying to launch/access spark

```16/08/17 15:36:04 INFO ExecutorRunner: Launch command: "/usr/lib/jvm/java-7-oracle/bin/java" "-cp" "/home/travis/build/apache/zeppelin/spark-1.6.1-bin-hadoop2.3/conf/:/home/travis/build/apache/zeppelin/spark-1.6.1-bin-hadoop2.3/lib/spark-assembly-1.6.1-hadoop2.3.0.jar:/home/travis/build/apache/zeppelin/spark-1.6.1-bin-hadoop2.3/lib/datanucleus-core-3.2.10.jar:/home/travis/build/apache/zeppelin/spark-1.6.1-bin-hadoop2.3/lib/datanucleus-api-jdo-3.2.6.jar:/home/travis/build/apache/zeppelin/spark-1.6.1-bin-hadoop2.3/lib/datanucleus-rdbms-3.2.9.jar" "-Xms1024M" "-Xmx1024M" "-Dspark.driver.port=47348" "-XX:MaxPermSize=256m" "org.apache.spark.executor.CoarseGrainedExecutorBackend" "--driver-url" "spark://CoarseGrainedScheduler@172.17.14.104:47348" "--executor-id" "0" "--hostname" "172.17.14.104" "--cores" "2" "--app-id" "app-20160817153604-0003" "--worker-url" "spark://Worker@172.17.14.104:41888"
16/08/17 15:36:40 INFO Worker: Asked to kill executor app-20160817153604-0003/0
16/08/17 15:36:40 INFO ExecutorRunner: Runner thread for executor app-20160817153604-0003/0 interrupted
16/08/17 15:36:40 INFO ExecutorRunner: Killing process!
16/08/17 15:36:41 INFO Worker: Executor app-20160817153604-0003/0 finished with state KILLED exitStatus 143
16/08/17 15:36:41 INFO Worker: Cleaning up local directories for application app-20160817153604-0003
16/08/17 15:36:41 INFO ExternalShuffleBlockResolver: Application app-20160817153604-0003 removed, cleanupLocalDirs = true

travis_time:end:17fafc60:start=1471448259153159015,finish=1471448259166402551,duration=13243536
```

Anything else I can do to get this merged?

bzz commented:
Sorry, been a busy week.  I think it needs to be rebased on latest master.

Few things on licensing:

There are two same **CDDL 1.1** texts now  `zeppelin-distribution/src/bin_license/LICENSE-jersey-2` and `zeppelin-distribution/src/bin_license/licenses/LICENSE-jersey-2` - I think we need to keep only the one inside `zeppelin-distribution/src/bin_license/licenses/`.

Please pardon me if my previous comment was confusing:
> I will look more into that, meanwhile could you please include full text of this license to zeppelin-distribution/src/bin_license/ as well?

What I meant was: we need to add full text of **CDDL 2.0** in the `zeppelin-distribution/src/bin_license/licenses/LICENSE-<component name>`

--------------

Looking a bit deeper, could you please double-check that all the dependencies mentioned in LICENSE file have correct licenses?

```
    (CDDL 2.0) javax.ws.rs-api (javax.ws.rs - https://jax-rs-spec.java.net/)
    (CDDL 2.0) validation-api (javax.validation - http://beanvalidation.org/)
    (CDDL 2.0) javax.annotation-api (javax.annotation - https://www.jcp.org/en/jsr/detail?id=250)
```

 -  https://jax-rs-spec.java.net/ - page mentiones only [CDDL 1.0](https://opensource.org/licenses/cddl1.php)
 -  http://beanvalidation.org/licensing/) - page mentiones only _Apache 2.0_ and JCP _click through license_
 - https://www.jcp.org/en/jsr/detail?id=250 the only mention of word `license` I could find is `JDL`

   > The Reference Implementation and associated TCK's will be made available from java.sun.com at no charge without support. J2EE and J2SE licensees will recieve support at no extra charge with an amendment to their active support agreement. Source code will be made available via the Java Distribution License (JDL).

And I can not find any mentions of CDDL 2.0 on https://opensource.org/licenses/alphabetical

bzz commented:
After the licenses clarification, I think we should be good to merge - I love the idea of having Jersey2.

\cc @Leemoonsoo @jongyoul for review as well

echarles commented:
Merged with latest master.

@bzz I don't know where I found CDDL 2.0 (sounded indeed strange when I was writing this some time ago). jax-rs is now CDDL 1.0, beanvalidation is now ASL2 and finally, excluded javax.annotation jar (classes are shipped in JRE).

bzz commented:
Thank you for addressing licensing issues! 

Let me look into that again and then merge to master, if there is no further discussion.

echarles commented:
@bzz pushed my local merge. Anything else that should be done before you merge?

bzz commented:
@echarles I'm terribly sorry for for such a huge communication delay in my side :\

Would you be willing to re-base it on the latest master so we could merge it?

jongyoul commented:
@echarles Hi, I saw your commit and looks better for me. At the first time, I don't understand what it is. Can you enable your travis to pass CI?

jongyoul commented:
In general, including Zeppelin, most of reviewers feel uncomfortable to upgrade some libraries even it's too old because no one guarantee all existing functions work perfectly - eg. Kafka had used Zookeeper 3.3.4 recently. Then, almost projects only upgrade their libraries when they found some big bugs or necessary features. In this case, when you introduce this upgrade feature, we don't need to upgrade its library because it worked well. But after upgrading jetty from 8 to 9, we found bugs and need to upgrade or fix some related versions. AFAIK, it needs later servlet version and it causes to upgrade jersey version. Still, I don't know why we keep cxf and jersey at the same time for a long time, but if your PR passes CI and works proper when reviewers test it, we can consider removing cxf and change jersey2 fully. Anyway, thanks for the contribution and I appreciate and encourage your contribution.

echarles commented:
Jenkins build is now green. @jongyoul  @bzz Any other inputs/questions on this PR ?

jongyoul commented:
@echarles Sorry, I haven't checked it correctly. Thanks for checking it again. I'll test it

sohaibiftikhar commented:
I was testing this branch. I had a scenario in which on using LDAPRealm the method not found exception was thrown as described in #2182 . In a similar scenario after building from this branch I got the following error when accessing /api/notebooks in the network requests.
`<body><h2>HTTP ERROR 405</h2>
<p>Problem accessing /api/login;jsessionid=gmhaugjgp4hg1l35w40eon2bp. Reason:
<pre>    Method Not Allowed</pre></p><hr><i><small>Powered by Jetty://</small></i><hr/>
</body>`

I am not sure if this is related to this PR will post back after further investigation however I do not know why /login is being called internally via a GET request. Any light on this would be very helpful.

Thanks!

jongyoul commented:
@echarles Any comment?

echarles commented:
It is working here with shiro-enabled (simple username password in shiro,ini, not the ldaprealm).
I would need to test with ldaprealm...
@sohaibiftikhar  is it working for your with the simple username/password in shiro.ini ?

sohaibiftikhar commented:
@echarles I tested with both simple and LdapGroupRealm and it was working. It is only with LdapRealm that I am facing an issue. I will post a sample `shiro.ini` for your reference. As I said I am still not sure why /login is being called internally. I tried debugging with the source code but due to my unfamiliarity with the codebase it has not been of much help.

```
[main]
 ldapRealm = org.apache.zeppelin.realm.LdapRealm
 ldapRealm.userDnTemplate = uid={0},ou=people,dc=my-company,dc=net
 ldapRealm.searchBase = dc=my-company,dc=net
 ldapRealm.userSearchBase = ou=people,dc=my-company,dc=net
 ldapRealm.groupSearchBase = ou=groups,dc=my-company,dc=net
 ldapRealm.contextFactory.url = ldaps://auth.mycompany.net:636
 ldapRealm.contextFactory.authenticationMechanism = simple
 ldapRealm.userObjectClass = posixAccount
 ldapRealm.groupObjectClass = posixGroup
 ldapRealm.authorizationEnabled = true
 ldapRealm.memberAttribute = memberUid
 ldapRealm.memberAttributeValueTemplate=uid={0},ou=people,dc=my-company,dc=net
 ldapRealm.rolesByGroup = USERS:admin
 ldapRealm.userSearchAttributeName = uid
 securityManager.realms = $ldapRealm
 sessionManager = org.apache.shiro.web.session.mgt.DefaultWebSessionManager
 shiro.loginUrl = /api/login

 [roles]
 admin = *

 [urls]
 /api/version = anon
 /** = authc
```

Eventually I will need group based permissions but this is just a simplified ini. Here is a screenshot of the network activity on calling api/notebooks.

[https://ibb.co/jLw9vk](url)

jongyoul commented:
I also had a similar problem. We need to handle it.

On Mon, May 8, 2017 at 11:31 PM, Sohaib Iftikhar <notifications@github.com>
wrote:

> @echarles <https://github.com/echarles> I tested with both simple and
> LdapGroupRealm and it was working. It is only with LdapRealm that I am
> facing an issue. I will post a sample shiro.ini for your reference. As I
> said I am still not sure why /login is being called internally. I tried
> debugging with the source code but due to my unfamiliarity with the
> codebase it has not been of much help.
>
> [main]
>  ldapRealm = org.apache.zeppelin.realm.LdapRealm
>  ldapRealm.userDnTemplate = uid={0},ou=people,dc=my-company,dc=net
>  ldapRealm.searchBase = dc=my-company,dc=net
>  ldapRealm.userSearchBase = ou=people,dc=my-company,dc=net
>  ldapRealm.groupSearchBase = ou=groups,dc=my-company,dc=net
>  ldapRealm.contextFactory.url = ldaps://auth.mycompany.net:636
>  ldapRealm.contextFactory.authenticationMechanism = simple
>  ldapRealm.userObjectClass = posixAccount
>  ldapRealm.groupObjectClass = posixGroup
>  ldapRealm.authorizationEnabled = true
>  ldapRealm.memberAttribute = memberUid
>  ldapRealm.memberAttributeValueTemplate=uid={0},ou=people,dc=my-company,dc=net
>  ldapRealm.rolesByGroup = USERS:admin
>  ldapRealm.userSearchAttributeName = uid
>  securityManager.realms = $ldapRealm
>  sessionManager = org.apache.shiro.web.session.mgt.DefaultWebSessionManager
>  shiro.loginUrl = /api/login
>
>  [roles]
>  admin = *
>
>  [urls]
>  /api/version = anon
>  /** = authc
>
> Eventually I will need group based permissions but this is just a
> simplified ini. Here is a screenshot of the network activity on calling
> api/notebooks.
>
> https://ibb.co/jLw9vk <http://url>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/zeppelin/pull/932#issuecomment-299883453>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADcfliiVLeqVHw60TWz93T3v-T0OdjSsks5r3ydHgaJpZM4Ip6YI>
> .
>



-- 
이종열, Jongyoul Lee, 李宗烈
http://madeng.net

sohaibiftikhar commented:
@jongyoul I'd be happy to look into it but would be better for me if you can provide a starting point to start looking at.

jongyoul commented:
Actually, I'm not sure I didn't dig into it. My current assumption is to support anonymous and authorized one

sohaibiftikhar commented:
@jongyoul Should I raise this as a separate issue because this seems unrelated to this PR? Apparently the ticket is not held when using LdapRealm while with LdapGroupRealm it is fine. However from my guess this happens at the Spring Security layer which makes it hard to debug for me.

sohaibiftikhar commented:
So I figured out what was happening. The problem was with my shiro.ini as I had not assigned a session manager.
`securityManager.sessionManager = $sessionManager`
After assigning this it works fine.

jongyoul commented:
@sohaibiftikhar Thanks for verifying it.

sohaibiftikhar commented:
@jongyoul I wish to open a dependent pull request for review. Is it possible to merge this branch if everything is working fine.

jongyoul commented:
@sohaibiftikhar Sure, I didn't test it yet as I had a talk in ApacheCon last Wednesday. I'll handle it soon.

sohaibiftikhar commented:
Hi @jongyoul 
Is there some way I can help push this to merge faster?

FRosner commented:
Hi all,

tested it and it works. I understood that https://github.com/apache/zeppelin/pull/2182 is on hold because this one is also going to fix it but in a better way? Can we move forward with this one?

Leemoonsoo commented:
Tested and LGTM.

Merge to master if no further discussions.

jongyoul commented:
Tested and LGTM. Sorry for the late reply. I'll merge it now.
